### PR TITLE
Add an analyzer that complains when using using model binding attributes with a minimal action delegate

### LINKED ISSUE #35088: Create an Analyzer to help users understand acceptable attributes for Minimal APIs
### Is your feature request related to a problem? Please describe.

**We would like to help the user understands which attributes are allowed for Minimal Actions during compile time.  see example below for Bind Attribute:** 

```csharp
  app.MapGet("/user", ([Bind("FirstName,LastName")] User user) => $"username {user.FirstName}");
```

### Describe the solution you'd like
Create an Analyzer to help the user understands which attributes are supported for Minimal APIs.

### Additional context
See above.

### PR DESCRIPTION
**Pull request by pranavkm**
* [x] Complain when using MVC binding attributes with minimal action

Fixes https://github.com/dotnet/aspnetcore/issues/35088

### CODE REVIEW

File: src/Framework/Analyzer/src/MinimalActions/DiagnosticDescriptors.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.AspNetCore.Analyzers.MinimalActions
+{
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("MicrosoftCodeAnalysisReleaseTracking", "RS2008:Enable analyzer release tracking")]
+    internal static class DiagnosticDescriptors
+    {
+        internal static readonly DiagnosticDescriptor DoNotUseModelBindingAttributesOnMinimalActionParameters = new(
+            "ASP0003",
+            "Do not use model binding attributes with minimal actions",
```
Comment (davidfowl) [inline src/Framework/Analyzer/src/MinimalActions/DiagnosticDescriptors.cs:14]: Lets remove the minimal action name.

File: src/Framework/Analyzer/test/MinimalActions/RouteAttributeMismatchTest.cs [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+    [Fact]
+    public async Task MinimalAction_SomeUnusedTokens_ProducesDiagnostics()
+    {
+        // Arrange
+        var source = TestSource.Read(@"
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Mvc;
+var webApp = WebApplication.Create();
+webApp.MapPost(/*MM*/""/{id:int}/{location:alpha}"", (int id, string loc) => {});
```
Comment (BrennanConroy) [inline src/Framework/Analyzer/test/MinimalActions/RouteAttributeMismatchTest.cs]: Can we have an analyzer also complain about `string loc`? This route is unusable with the non-bindable `loc` parameter.

File: src/Framework/Analyzer/test/MinimalActions/RouteAttributeMismatchTest.cs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+    [Fact]
+    public async Task MinimalAction_FromRouteParameterWithMatchingToken_Works()
+    {
+        // Arrange
+        var source = TestSource.Read(@"
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Mvc;
+var webApp = WebApplication.Create();
+webApp.MapPost(/*MM*/""/{id:int}"", ([FromRoute] int id, string loc) => {});
```
Comment (BrennanConroy) [inline src/Framework/Analyzer/test/MinimalActions/RouteAttributeMismatchTest.cs]: Same

File: src/Framework/Analyzer/src/MinimalActions/DiagnosticDescriptors.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+namespace Microsoft.AspNetCore.Analyzers.MinimalActions
+{
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("MicrosoftCodeAnalysisReleaseTracking", "RS2008:Enable analyzer release tracking")]
+    internal static class DiagnosticDescriptors
+    {
+        internal static readonly DiagnosticDescriptor DoNotUseModelBindingAttributesOnMinimalActionParameters = new(
+            "ASP0003",
+            "Do not use model binding attributes with Map actions",
+            "{0} should not be specified for a {1} delegate parameter",
```
Comment (pranavkm) [inline src/Framework/Analyzer/src/MinimalActions/DiagnosticDescriptors.cs:14]: > Can we point them at the correct solution?

Could you recommend what would be?

Comment (pranavkm) [reply src/Framework/Analyzer/src/MinimalActions/DiagnosticDescriptors.cs:14]: /cc @BrennanConroy

Comment (BrennanConroy) [reply src/Framework/Analyzer/src/MinimalActions/DiagnosticDescriptors.cs:14]: I don't know, I haven't used map actions much

Comment (pranavkm) [reply src/Framework/Analyzer/src/MinimalActions/DiagnosticDescriptors.cs:14]: There's a few things model binding can do - say that you want to bind a loose collection of key value pairs from route / query / form (no support), say you want to bind a subset, or say that you want to bind some value with a different name. It's not trivial to figure out what the user's intent is, to give people something very prescriptive.

File: src/Framework/Analyzer/src/MinimalActions/DiagnosticDescriptors.cs [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+            "Usage",
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true,
+            helpLinkUri: "https://aka.ms/minimal-action/analyzer");
```
Comment (BrennanConroy) [inline src/Framework/Analyzer/src/MinimalActions/DiagnosticDescriptors.cs:18]: What is this link going to point to? And should we remove minimal from it

**Review by BrennanConroy — APPROVED** [review]
Is there a way to do full end to end tests? Otherwise we're just hoping the packaging works correctly and we don't accidentally break it in the future.

File: src/Framework/Analyzer/src/MinimalActions/DiagnosticDescriptors.cs (hunk 5 above)
Comment (pranavkm) [reply src/Framework/Analyzer/src/MinimalActions/DiagnosticDescriptors.cs:18]:
```suggestion
            helpLinkUri: "https://aka.ms/aspnet-app/analyzers");
```

How about this? The plan is to have it point to some docs like the ones NetAnalyzers does.

Comment (BrennanConroy) [reply src/Framework/Analyzer/src/MinimalActions/DiagnosticDescriptors.cs:18]: Sounds good, and we'll eventually have more analyzers not minimal specific that can go there too.

File: eng/Dependencies.props [hunk 6]
```diff
@@ -191,6 +191,7 @@ and are generated based on the last package release.
     <LatestPackageReference Include="Serilog.Extensions.Logging" />
     <LatestPackageReference Include="Serilog.Sinks.File" />
     <LatestPackageReference Include="StackExchange.Redis" />
+    <LatestPackageReference Include="System.Memory" />
```
Comment (captainsafia) [inline eng/Dependencies.props:194]: What is this used for?

File: src/Analyzers/Microsoft.AspNetCore.Analyzer.Testing/src/TestSource.cs [hunk 7]
```diff
@@ -26,9 +26,15 @@ public static TestSource Read(string rawSource)
             for (var i = 0; i < lines.Length; i++)
             {
                 var line = lines[i];
-                var markerStartIndex = line.IndexOf(MarkerStart, StringComparison.Ordinal);
-                if (markerStartIndex != -1)
+
+                while (true)
```
Comment (captainsafia) [inline src/Analyzers/Microsoft.AspNetCore.Analyzer.Testing/src/TestSource.cs:30]: Why do we use the while-loop/break pattern here now?

Comment (pranavkm) [reply src/Analyzers/Microsoft.AspNetCore.Analyzer.Testing/src/TestSource.cs:30]: A single line can have more than one diagnostics in it. The latter ones were previously being ignored.

**Pull request merged**